### ISSUE TITLE
decide which versions of java to target

### ISSUE THREAD

**Issue by lbergelson** [samtools/htsjdk-next-beta#18]

**Comment by lbergelson** [comment]
My opinion is that we can't drop support for java 8 in any predictable time frame despite it's end of life.  GATK has a dependency on spark, and spark has no concrete plans to support any java past 8 so far.  There are substantial difficulties moving many libraries from 8 -> 9+ because of changes to methods that while technically unsupported were widely used for libraries that do low level memory management.  

It would be good to be compatible with java 9+ so that downstream tools can upgrade. 

I don't know if it's possible to use features of java 9/10/11 in the code and then build libraries that support running under java 8, but if that's possible it seems like it might be a good idea.

**Comment by lbergelson** [comment]
[This issue](https://issues.apache.org/jira/browse/SPARK-24417?jql=project%20%3D%20SPARK%20AND%20text%20~%20%22support%20java%2011%22) tracks spark development towards running on java 11.

**Comment by magicDGS** [comment]
I don't think that we should drop support for java8 yet, but it will be nice to be compatible with later versions (at least LTS). As we build on top of openjdk, we can follow the roadmap of one of the build providers. For example, [AdoptOpenJDK](https://adoptopenjdk.net/support.html) will support java8 until September 2022 to allow people to transition to java11 and onwards. I think that the test matrix in travis should include:

- openjdk8 to test as an target/source version 
- openjdk11 once it is realease, to support users adopting java11
- from there onwards, we can build with the non-LTS that are released to check if we are not breaking anything with our build

Maybe we can also have a branch for java11 if we are interested in having new features, or directly start with java11 before release to move the community to the latest version (as far as Hadoop/Spark supports java11).

**Comment by jrobinso** [comment]
We,  where we primarily means David Eby,  have been working on JDK support for 9, 10, and 11 for some time.    I'm sure you are aware this is a non-trivial port.   

We have Java 10 released now,  but 11 poses a significant problem for us as it does not include JAXB.   You can install it as a separate library,  but the library is GPL 2 licensed.    We are not going to force GPL on IGV users so that's not an option for us.   We can remove the use of JAXB for IGV code,  but unfortunately the htkjdk uses it.   We might have to fork htsjdk and remove the use of JAXB.  Hopefully its not used for anything critical.

So you will face this question if you attempt a JDK 11 release as well,  do you keep JAXB and change your license to GPL or remove the dependency on it?   This will of course affect what we do as well.

**Comment by davideby** [comment]
FWIW, JAXB is just our latest headache.  We've had issues big & small all along the way in going from 8 to 9 to 10 and now working on 11.  That could be because IGV is a GUI app with lots of dependencies.  You seem to have fewer so maybe you'll be in better shape.  You won't really know until you start.

Not meaning to hijack this into an IGV thread, just adding some perspective...

**Comment by lbergelson** [comment]
Thank you for the information.  It seems likely that JAXB could be removed from htsjdk 3.  It would be harder to remove it from the existing htsjdk because that would break the people depending on it.  From what I saw it looked like [JAXB](https://javaee.github.io/jaxb-v2/) is available as GPL2 *with the class path exception*.  My understanding was that that allowed you to include the library with your application without requiring that your application had any particular license.  Maybe I'm misunderstanding that though?  

I'm hoping that we will have fewer issues going to java 9+ because we have few dependencies, but as you say, it's hard to know.

**Comment by droazen** [comment]
Chiming in: I'd hope that we'd maintain support for OpenJDK 8 for quite some time, given that it's officially supported through ~2022, and given that downstream projects like GATK are currently blocked from upgrading beyond it.

**Comment by jrobinso** [comment]
@lbergelson I think you are right about the classpath exception.  Sorry for the false alarm.  There is another section on the GNU site that explicitly uses including a Java "library" as making a combined work,  with no mention of CP, that threw me off.

This Java release is unlike any before it.   For example, if you go to the Oracle download page Java 8 is still the recommended download, even though 9 & 10 have been released for a while and are nearing their EOL.   Also, it seems likely that Java 11 from Oracle will not be free.

**Comment by magicDGS** [comment]
Sorry for the delay on this: I am not saying to remove support to java8 by now but to be source-compatible with java 11 - that means, do not use JAXB (there are few usages in the previous code - quickly checked with [this](https://github.com/samtools/htsjdk/search?q=jaxb&unscoped_q=jaxb&type=Code) and [this](https://github.com/samtools/htsjdk/search?q=xml&unscoped_q=xml&type=Code). In case of required for some downstream users, they can add their own wrappers to annotate the methods that they need, as we are trying to design a proper interface for the main usages of JAXB for serialization.

The main idea of the issue was to discuss if we add to the travis matrix a check for building with LTS-versions only (11 by now) or also add entries to the matrix for non-LTS (9-10, etc). Later, we can figure out how we can evolve and stop supporting java8 as the community evolve (see [this post](https://medium.com/criciumadev/its-time-migrating-to-java-11-5eb3868354f9) for the migration guide and some of the major points to upgrade). I expect (maybe I am too optimistic) that even before the first release of htsjdk3, java11 will be already adopted by most of the major blockers in downstream clients (like Picard/GATK) - in that case, it will be easier to transition the library if we are already source-compatible by removing dependencies on Oracle JDK removed APIs (and stick to openJDK).
